plutus-metatheory/src/Untyped/CEK.lagda.md

```diff
@@ -405,30 +417,48 @@ BUILTIN' b {pt = pt} {pa = pa} bt with trans (sym (+-identityʳ _)) (∔2+ pt) |
-step (s ; ρ ▻ (t · u))   = (s , -· ρ u) ; ρ ▻ t
-step (s ; ρ ▻ force t)   = (s , force-) ; ρ ▻ t
-step (s ; ρ ▻ delay t)   = s ◅ V-delay ρ t
-step (s ; ρ ▻ con (tmCon t c)) = s ◅ V-con t c
-step (s ; ρ ▻ builtin b) = s ◅ ival b
-step (s ; ρ ▻ error)     = ◆
-step (ε ◅ v)             = □ v
-step ((s , -· ρ u) ◅ v)  = (s , (v ·-)) ; ρ ▻ u
-step ((s , (V-ƛ ρ t ·-)) ◅ v) = s ; ρ ∷ v ▻ t
-step ((s , (V-I⇒ b {am = 0} bapp ·-)) ◅ v) =
-  s ; [] ▻ BUILTIN' b (app bapp v)
-step ((s , (V-I⇒ b {am = suc _} bapp ·-)) ◅ v) =
-  s ◅ V-I b (app bapp v)
-step ((s , (V-con _ _   ·-)) ◅ v) = ◆ -- constant in function position
-step ((s , (V-delay _ _ ·-)) ◅ v) = ◆ -- delay in function position
-step ((s , (V-IΠ b bapp ·-)) ◅ v) = ◆ -- delayed builtin in function position
-step ((s , force-) ◅ V-IΠ b bapp) = s ◅ V-I b (app⋆ bapp)
-step ((s , force-) ◅ V-delay ρ t) = step (s ; ρ ▻ t)
-step ((s , force-) ◅ V-ƛ _ _)     = ◆ -- lambda in delay position
-step ((s , force-) ◅ V-con _ _)   = ◆ -- constant in delay position
-step ((s , force-) ◅ V-I⇒ b bapp) = ◆ -- function in delay position
+step (s ; ρ ▻ ` x)               = s ◅ lookup ρ x
+step (s ; ρ ▻ ƛ t)               = s ◅ V-ƛ ρ t
+step (s ; ρ ▻ (t · u))           = (s , -· ρ u) ; ρ ▻ t
+step (s ; ρ ▻ force t)           = (s , force-) ; ρ ▻ t
+step (s ; ρ ▻ delay t)           = s ◅ V-delay ρ t
+step (s ; ρ ▻ con (tmCon t c))   = s ◅ V-con t c
+step (s ; ρ ▻ constr i [])       = s ◅ V-constr i []
+step (s ; ρ ▻ constr i (x ∷ xs)) = (s , constr- i [] ρ xs); ρ ▻ x
+step (s ; ρ ▻ case t ts)         = (s , case- ρ ts) ; ρ ▻ t
+step (s ; ρ ▻ builtin b)         = s ◅ ival b
+step (s ; ρ ▻ error)             = ◆
+
+step (ε ◅ v)                               = □ v
+step ((s , -· ρ u) ◅ v)                    = (s , (v ·-)) ; ρ ▻ u
+step ((s , -·v v) ◅ vf)                    = (s , vf ·-) ◅ v
+step ((s , (V-ƛ ρ t ·-)) ◅ v)              = s ; ρ ∷ v ▻ t
+step ((s , (V-con _ _   ·-)) ◅ v)          = ◆ -- constant in function position
+step ((s , (V-delay _ _ ·-)) ◅ v)          = ◆ -- delay in function position
+step ((s , (V-IΠ b bapp ·-)) ◅ v)          = ◆ -- delayed builtin in function position
+step ((s , (V-constr i vs ·-)) ◅ v)        = ◆ -- SOP in function position
+step ((s , force-) ◅ V-IΠ b bapp)          = s ◅ V-I b (app⋆ bapp)
+step ((s , force-) ◅ V-delay ρ t)          = step (s ; ρ ▻ t)
+step ((s , force-) ◅ V-ƛ _ _)              = ◆ -- lambda in delay position
+step ((s , force-) ◅ V-con _ _)            = ◆ -- constant in delay position
+step ((s , force-) ◅ V-I⇒ b bapp)          = ◆ -- function in delay position
+step ((s , force-) ◅ V-constr i vs)        = ◆ -- SOP in delay position
+step ((s , constr- i vs ρ []) ◅ v)         = s ◅ V-constr i (v ∷ vs)
+step ((s , constr- i vs ρ (x ∷ ts)) ◅ v)   = (s , constr- i (v ∷ vs) ρ ts); ρ ▻ x
+step ((s , case- ρ ts) ◅ V-constr i vs)    = maybe (extendStack s vs ; ρ ▻_) ◆ (lookup? i ts)
+step ((s , case- _ _ ) ◅ v)                = ◆ -- case of not a SOP
```

Comment:
Haven't read this function before, so slightly off-topic, but I'm not sure what your approach is here: sometimes you list out all the "wrong" cases explicitly (like for all the cases for function application and forcing), sometimes you use a fall-through pattern like here. Is there a reason for the difference?

Reply:
I don't really have an answer. I guess the previous definitions were listed explicitly to make sure that all the good cases are being handled? I should probably follow the existing style and list all cases...